opentreemap/treemap/audit.py

```diff
@@ -1268,11 +1268,14 @@ def _deserialize_value(self, value):
                 udfd_pk = int(self.model[4:])
                 udf_def = next((udfd for udfd in udfds if udfd.pk == udfd_pk),
                                None)
-                datatype = udf_def.datatype_by_field[field_name]
+                if udf_def is not None:
+                    datatype = udf_def.datatype_by_field[field_name]
             else:
                 udf_def = next((udfd for udfd in udfds
-                                if udfd.name == field_name), None)
-                datatype = udf_def.datatype_dict
+                                if udfd.name == field_name
+                                and udfd.model_type == self.model), None)
+                if udf_def is not None:
+                    datatype = udf_def.datatype_dict
             if udf_def is not None:
                 return udf_def.clean_value(value, datatype)
```

Comment:
I think it is better not to have conditionally defined variables that rely on a boolean to not trigger a `NameError`. If some aspect of the spec changes such that a developer changes that boolean such that it may be `True` sometimes (but not always) even when `udf_def` is `None`, then they may commit a `NameError` that they failed to produce in development.

Safer is something like:
```
e = Exception('Cannot format a UDF audit whose definition has been'
              ' deleted! This audit should be deleted as well')
if self.model.startswith('udf:'):
    udfd_pk = int(self.model[4:])
    udf_def = next((udfd for udfd in udfds if udfd.pk == udfd_pk),
                   None)
    if udf_def is not None:
        datatype = udf_def.datatype_by_field[field_name]
    else:
        raise e
else:
    udf_def = next((udfd for udfd in udfds
                    if udfd.name == field_name
                    and udfd.model_type == self.model), None)
    if udf_def is not None:
        datatype = udf_def.datatype_dict
    else:
        raise e
return udf_def.clean_value(value, datatype)
```